pkg/logger/logger.go

```diff
@@ -55,3 +56,47 @@ func logLevelFromEnvironment() LogLevel {
 	}
 	return LogInfo
 }
+
+type (
+	TimeFunc = func() time.Time
+
+	// Options can be used to configure a new logger.
+	Options struct {
+		// Name of the subsystem to prefix logs with
+		Name string
+
+		// The threshold for the logger. Anything less severe is supressed
+		Level LogLevel
+
+		// Where to write the logs to. Defaults to os.Stderr if nil
+		Output io.Writer
+
+		// The time format to use instead of the default
+		TimeFormat string
+
+		// A function which is called to get the time object that is formatted using `TimeFormat`
+		TimeFnc TimeFunc
+
+		// The format in which logs will be formatted. (eg: text/json)
+		LogFormat string
+
+		// The file to write to.
+		LogFile string
+	}
+)
+
+// NewLogger is a factory for selecting a logger based on options
+func NewLogger(opts *Options) (logger Logger, out *os.File, err error) {
```

Comment:
opts could be of type Options, it avoids checking for null pointers

Reply:
Depends really, can use the null check to use default options in the future, and to avoid any default struct values for implicit data types (string = "", int = 0 , bool = false etc)

Reply:
fair point, in other parts of the codebase there are defaultOptions which can be modified to customize them

Reply:
ack